hubble-relay: add Dockerfile and make target to build hubble-relay image

rolinh opened the pull request.

rolinh committed:
hubble-relay: add Dockerfile and make target to build hubble-relay image

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

aanm reviewed (changes requested):
Please update the release process documentation stating that this image also needs to be taken care of when doing a release https://github.com/cilium/cilium/blob/3ca38b3f182da4611a12d43621f284e5e8d12960/Documentation/contributing/release/stable.rst#L179-L180

rolinh commented on hubble-relay.Dockerfile:
```diff
@@ -0,0 +1,24 @@
+COPY --from=builder /go/src/github.com/cilium/cilium/hubble-relay/hubble-relay /usr/bin/hubble-relay
+COPY --from=certs /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/ca-certificates.crt
+# cilium-sysdump expects gops to be installed in /bin
+COPY --from=gops /go/bin/gops /bin
```
NOTE: gops is not expected to work just yet (support needs to be added to `hubble-relay serve`).

aanm replied:
@Rolinh not only that but I think the binary itself won't run as well because it's not statically linked.

rolinh replied:
This will be implemented and tested as part of #11227.

gandro reviewed (approved):
Not super familiar with the release process, but the Dockerfile itself looks good to me!

tklauser commented on Makefile:
```diff
@@ -285,6 +285,12 @@ docker-cilium-builder-manifest:
+	$(CONTAINER_ENGINE_FULL) build -f hubble-relay.Dockerfile -t "cilium/hubble-relay:$(DOCKER_IMAGE_TAG)" .
+	$(CONTAINER_ENGINE_FULL) tag cilium/hubble-relay:$(DOCKER_IMAGE_TAG) cilium/hubble-relay:$(DOCKER_IMAGE_TAG)-${GOARCH}
+	$(QUIET)echo "Push like this when ready:"
+	$(QUIET)echo "docker push cilium/hubble-relay:$(DOCKER_IMAGE_TAG)-${GOARCH}"
```
Nit: use `${CONTAINER_ENGINE}` instead of `docker` in the push instructions:

```suggestion
	$(QUIET)echo "
	${CONTAINER_ENGINE} push cilium/hubble-relay:$(DOCKER_IMAGE_TAG)-${GOARCH}"
```

I think this might also be wrong in some of the other `docker-*-image` targets where this was probably copied from.

rolinh replied:
It was indeed copied from the `docker-operator-image` target. I can update both targets if you want.

tklauser replied:
That'd be great, thanks.

tklauser commented on hubble-relay.Dockerfile:
```diff
@@ -0,0 +1,24 @@
+FROM docker.io/library/alpine:3.11 as certs
+RUN apk --update add ca-certificates
+
+FROM docker.io/library/golang:1.14.2-alpine3.11 as gops
```
I think the `update-golang-dockerfiles` rule in the main `Makefile` (which is used to update the Go version in all Dockerfiles) will not match this correctly, i.e. it will strip the `-alpine3.11`.

rolinh replied:
OK, I'll update this section to be aligned with what we have in `contrib/packaging/docker/Dockerfile.runtime` so that the `update-golang-dockerfiles` target will continue to work without any modification.

tklauser replied:
Perfect, thanks.

rolinh committed:
make: update some make targets to use `${CONTAINER_ENGINE}`

The "${CONTAINER_ENGINE}" variable should be used instead of "docker"
within make targets. This commits fixes 2 instances where "docker" was
used instead of the variable.

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>
